src/main/kotlin/org/jmailen/gradle/kotlinter/KotlinterPlugin.kt

```diff
@@ -20,146 +21,117 @@ import org.jmailen.gradle.kotlinter.tasks.LintTask
             it.group = "formatting"
             it.description = "Runs lint on the Kotlin source files."
-            it.dependsOn(lintTasks)
+        }.also {
+            tasks.named("check").dependsOn(it)
         }
 
-        project.tasks.getByName("check") {
-            it.dependsOn(lintKotlin)
+    private fun Project.registerParentFormatTask() =
+        tasks.register("formatKotlin") {
+            it.group = "formatting"
+            it.description = "Formats the Kotlin source files."
         }
-    }
 }
 
-typealias SourceSetAction = (String, Set<File>) -> Unit
+internal typealias SourceSetAction = (String, Provider<FileTree>) -> Unit
 
 interface SourceSetResolver {
     fun applyToAll(project: Project, action: SourceSetAction)
 }
 
-object KotlinJvmSourceSetResolver : SourceSetResolver {
+internal object KotlinJvmSourceSetResolver : SourceSetResolver {
 
     override fun applyToAll(project: Project, action: SourceSetAction) {
         getSourceSets(project).all { sourceSet ->
-            getKotlinFiles(sourceSet)?.let { files ->
-                action(getSourceSetName(sourceSet)!!.id, files)
+            sourceSet.kotlin?.let { directorySet ->
+                action(directorySet.name.id, project.provider { directorySet })
             }
         }
     }
 
     private fun getSourceSets(project: Project): SourceSetContainer =
         project.convention.getPlugin(JavaPluginConvention::class.java).sourceSets
 
-    private fun getSourceSetName(sourceSet: SourceSet) = sourceSet.kotlin?.name
-
-    private fun getKotlinFiles(sourceSet: SourceSet) = sourceSet.kotlin?.files
-
     private val SourceSet.kotlin: SourceDirectorySet?
         get() = ((getConvention("kotlin") ?: getConvention("kotlin2js")) as? KotlinSourceSet)?.kotlin
 
     private fun SourceSet.getConvention(name: String): Any? =
         (this as HasConvention).convention.plugins[name]
 }
 
-object AndroidSourceSetResolver : SourceSetResolver {
+internal object AndroidSourceSetResolver : SourceSetResolver {
 
     override fun applyToAll(project: Project, action: SourceSetAction) {
         val android = project.extensions.findByName("android")
         (android as? BaseExtension)?.let {
             it.sourceSets.all { sourceSet ->
                 val id = sourceSet.name.id
-                val files = getKotlinFiles(project, sourceSet)
-                if (files.isNotEmpty()) {
-                    action(id, files)
-                }
+                action(id, project.provider { getKotlinFiles(project, sourceSet) })
             }
         }
     }
 
     private fun getKotlinFiles(project: Project, sourceSet: AndroidSourceSet) = sourceSet.java.srcDirs.map { dir ->
-        project.fileTree(dir) {
-            it.include("**/*.kt")
-        }
+        project.fileTree(dir) { it.include("**/*.kt") }
     }.reduce { merged: FileTree, tree ->
-        merged.plus(tree)
-    }.files
+        merged + tree
+    }
 }
 
-val String.id
+internal val String.id
```

Comment:
I let myself to set visibility of those to `internal` to avoid exposing those as that was something I've noticed when using `Kotlinter` plugin. I assume those aren't meant to be used outside the library, right?